File: 1593-split-a-string-into-the-max-number-of-unique-substrings/1593-split-a-string-into-the-max-number-of-unique-substrings.cpp
class Solution {
public:
    
    void dfs(string& s,unordered_set<string>& map, int& ans , int start){
        
        if(start == s.length()) {
            int ss = map.size();
            ans = max(ans,ss);
            return;
        }
        
        string curr = "";
        
        for(int i = start ; i< s.length();i++){
             curr +=s[i];
            if(map.find(curr) == map.end()){
                map.insert(curr);        
                dfs(s,map,ans,i+1);
                map.erase(curr);
            }
        }
    }
    
    int maxUniqueSplit(string s) {
        unordered_set<string> map;
        int ans = 0;
        dfs(s,map,ans,0);
        return ans;
        
    } 
};

File: 1593-split-a-string-into-the-max-number-of-unique-substrings/1593-split-a-string-into-the-max-number-of-unique-substrings.java
class Solution {
   
   int max = 0;
   
   public int maxUniqueSplit(String s) {
       backtracking(s, 0, new HashSet<>());
       return max;
   }
   
   private void backtracking(String s, int start, Set<String> set) {
       if (start == s.length()) {
           max = Math.max(max, set.size());
       } else {
           for (int i = start + 1; i <= s.length(); i++) {
               String string = s.substring(start, i);
               if (set.add(string)) {
                   backtracking(s, i, set);
                   set.remove(string);
               }
           }
       }
   }
}